libeos-parental-controls/app-filter.h

```diff
@@ -113,4 +113,72 @@ void          epc_get_app_filter_async  (GDBusConnection      *connection,
 EpcAppFilter *epc_get_app_filter_finish (GAsyncResult         *result,
                                          GError              **error);
 
+void          epc_set_app_filter_async  (GDBusConnection      *connection,
+                                         uid_t                 user_id,
+                                         EpcAppFilter         *app_filter,
+                                         gboolean              allow_interactive_authorization,
+                                         GCancellable         *cancellable,
+                                         GAsyncReadyCallback   callback,
+                                         gpointer              user_data);
+gboolean      epc_set_app_filter_finish (GAsyncResult         *result,
+                                         GError              **error);
+
+/**
+ * EpcAppFilterBuilder:
+ *
+ * #EpcAppFilterBuilder is a stack-allocated mutable structure used to build an
+ * #EpcAppFilter instance. Use epc_app_filter_builder_init(), various method
+ * calls to set properties of the app filter, and then
+ * epc_app_filter_builder_end(), to construct an #EpcAppFilter.
+ *
+ * Since: 0.1.0
+ */
+typedef struct
+{
+  /*< private >*/
+  gpointer p0;
+  gpointer p1;
+} EpcAppFilterBuilder;
+
+GType epc_app_filter_builder_get_type (void);
+
+/**
+ * EPC_APP_FILTER_BUILDER_INIT:
+ *
+ * Initialise a stack-allocated #EpcAppFilterBuilder instance at declaration
+ * time.
+ *
+ * This is typically used with g_auto():
+ * |[
+ * g_auto(EpcAppFilterBuilder) builder = EPC_APP_FILTER_BUILDER_INIT ();
+ * ]|
+ *
+ * Since: 0.1.0
+ */
+#define EPC_APP_FILTER_BUILDER_INIT() \
+  { \
+    g_ptr_array_new_with_free_func (g_free), \
+    g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL) \
+  }
+
+void epc_app_filter_builder_init  (EpcAppFilterBuilder *builder);
+void epc_app_filter_builder_clear (EpcAppFilterBuilder *builder);
+
+G_DEFINE_AUTO_CLEANUP_CLEAR_FUNC (EpcAppFilterBuilder,
+                                  epc_app_filter_builder_clear)
+
+EpcAppFilterBuilder *epc_app_filter_builder_new  (void);
+EpcAppFilterBuilder *epc_app_filter_builder_copy (EpcAppFilterBuilder *builder);
+void                 epc_app_filter_builder_free (EpcAppFilterBuilder *builder);
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (EpcAppFilterBuilder, epc_app_filter_builder_free)
+
+EpcAppFilter *epc_app_filter_builder_end (EpcAppFilterBuilder *builder);
+
+void epc_app_filter_builder_blacklist_path (EpcAppFilterBuilder   *builder,
+                                            const gchar           *path);
```

Comment:
How will this API work if it comes to pass that we want to add support for using a whitelist rather than a blacklist? This method would have to fail, right, because you can't mix the two.

(But then, because the struct layout is essentially public and has no padding, I suppose the answer is “this API will not support that”.)

Reply:
I’ve added some struct padding (thanks). At the moment, the answer is definitely ‘whitelist is not supported’ though. We should be free to change the API in future, though, since it’s only consumed by Endless code.

Reply:
> What's the point in struct padding if we are assuming that we're free to change the API because this is only consumed by Endless code?

I have in my head that we can change the API additively (so, actually, no API breaks) by adding a new `epc_app_filter_builder_whitelist_path()` method, with usage restrictions such that it’s a programmer error to use both it and the blacklist method. Essentially that would mean the whitelist/blacklist boolean gets set implicitly by the programmer’s choice of method to call.

> I guess the essence of what I'm asking is, why is the list stored as `(bas)` when the boolean is always `false` meaning blacklist. And the answer is, because while we are free to change the API without worrying about backwards compatibility, changing the format of the data stored on users' computers is rather harder, so it's worth forward-proofing that.

Yes, exactly.